packages/experiment-tag/src/experiment.ts

```diff
@@ -4,461 +4,654 @@ import {
+  private previousUrl: string | undefined = undefined;
+  // Cache to track exposure for the current URL, should be cleared on URL change
+  private urlExposureCache: Record<string, Record<string, string | undefined>> =
+    {};
+  private flagVariantMap: Record<string, Record<string, Variant>> = {};
+  private localFlagKeys: string[] = [];
+  private remoteFlagKeys: string[] = [];
+  private isRemoteBlocking = false;
+  private customRedirectHandler: ((url: string) => void) | undefined;
+
+  constructor(
+    apiKey: string,
+    initialFlags: string,
+    config: WebExperimentConfig = {},
+  ) {
+    this.apiKey = apiKey;
+    this.initialFlags = JSON.parse(initialFlags);
+    // merge config with defaults and experimentConfig (if provided)
+    this.config = {
+      ...Defaults,
+      ...config,
+      ...(this.globalScope?.experimentConfig ?? {}),
+    };
+    if (!isLocalStorageAvailable() || !this.globalScope) {
+      return;
+    }
 
-  // if web_exp_id does not exist:
-  // 1. if device_id exists, migrate device_id to web_exp_id and remove device_id
-  // 2. if device_id does not exist, create a new web_exp_id
-  // 3. if both device_id and web_exp_id exist, remove device_id
-  if (!user.web_exp_id) {
-    user.web_exp_id = user.device_id || UUID();
-    delete user.device_id;
-    globalScope.localStorage.setItem(
-      experimentStorageName,
-      JSON.stringify(user),
-    );
-  } else if (user.web_exp_id && user.device_id) {
-    delete user.device_id;
-    globalScope.localStorage.setItem(
-      experimentStorageName,
-      JSON.stringify(user),
-    );
-  }
+    if (this.globalScope?.webExperiment) {
+      return;
+    }
 
-  const urlParams = getUrlParams();
-  // if in visual edit mode, remove the query param
-  if (urlParams['VISUAL_EDITOR']) {
-    globalScope.history.replaceState(
-      {},
-      '',
-      removeQueryParams(globalScope.location.href, ['VISUAL_EDITOR']),
-    );
-    return;
-  }
+    this.globalScope.webExperiment = this;
+
+    const urlParams = getUrlParams();
 
-  let isRemoteBlocking = false;
-  const remoteFlagKeys: Set<string> = new Set();
-  const localFlagKeys: Set<string> = new Set();
-  const parsedFlags = JSON.parse(initialFlags);
-
-  parsedFlags.forEach((flag: EvaluationFlag) => {
-    const { key, variants, segments, metadata = {} } = flag;
-
-    // Force variant if in preview mode
-    if (
-      urlParams['PREVIEW'] &&
-      key in urlParams &&
-      urlParams[key] in variants
-    ) {
-      // Remove preview-related query parameters from the URL
-      globalScope.history.replaceState(
+    // if in visual edit mode, remove the query param
+    if (urlParams['VISUAL_EDITOR']) {
+      WindowMessenger.setup();
+      this.globalScope.history.replaceState(
         {},
         '',
-        removeQueryParams(globalScope.location.href, ['PREVIEW', key]),
+        removeQueryParams(this.globalScope.location.href, ['VISUAL_EDITOR']),
       );
+      return;
+    }
```

Comment:
I think I mislead you when we met about this. We should have most of this in the start() function. 

The constructor should just initialize all the required member fields. I.e. we should not be returning early here.